histomicstk/features/FeatureExtraction.py

```diff
@@ -73,374 +78,98 @@ def FeatureExtraction(Label, In, Ic, K=128, Fs=6, Delta=8):
-        # add centroids
-        CentroidX = np.append(CentroidX, region.centroid[0])
-        CentroidY = np.append(CentroidY, region.centroid[1])
-        # add morphometry features
-        Area = np.append(Area, region.area)
-        Perimeter = np.append(Perimeter, region.perimeter)
-        Eccentricity = np.append(Eccentricity, region.eccentricity)
-        numerator = 4 * math.pi * region.area
-        denominator = math.pow(region.perimeter, 2)
-        Circularity = np.append(
-            Circularity,
-            numerator / denominator if denominator else 0
-        )
-        MajorAxisLength = np.append(MajorAxisLength, region.major_axis_length)
-        MinorAxisLength = np.append(MinorAxisLength, region.minor_axis_length)
-        Extent = np.append(Extent, region.extent)
-        Solidity = np.append(Solidity, region.solidity)
         # get bounds of dilated nucleus
         min_row, max_row, min_col, max_col = \
-            GetBounds(region.bbox, Delta, size_x, size_y)
+            _GetBounds(regions[i].bbox, Delta, size_x, size_y)
+
         # grab nucleus mask
         Nucleus = (
-            Label[min_row:max_row, min_col:max_col] == region.label
-        ).astype(np.uint8)
-        # find nucleus boundaries
-        Bounds = np.argwhere(
-            find_boundaries(Nucleus, mode="inner").astype(np.uint8) == 1
-        )
-        # calculate and add FSDs
-        FSDGroup[regionIdx, :] = FSDs(
-            Bounds[:, 0], Bounds[:, 1],
-            K, Interval
-        )
+            Label[min_row:max_row, min_col:max_col] == regions[i].label
+        ).astype(np.bool)
+
         # generate object coords for nuclei and cytoplasmic regions
-        Nuclei[regionIdx] = region.coords
+        Nuclei = regions[i].coords
+
+        # compute Texture, Gradient, Intensity features
+        HematoxylinGradientGroup[i, :] = \
+            ComputeGradientFeatures(In, Nuclei, diffGn, BW_cannyn)
+
         # get mask for all nuclei in neighborhood
         Mask = (
             Label[min_row:max_row, min_col:max_col] > 0
         ).astype(np.uint8)
+
         # remove nucleus region from cytoplasm+nucleus mask
         cytoplasm = (
-            np.logical_xor(Mask, dilation(Nucleus, Disk))
-        ).astype(np.uint8)
-        # get list of cytoplasm pixels
-        Cytoplasms[regionIdx] = GetPixCoords(
-            cytoplasm, min_row, min_col)
-        regionIdx = regionIdx + 1
-
-    # calculate hematoxlyin features, capture feature names
-    HematoxylinIntensityGroup = IntensityFeatureGroup(In, Nuclei)
-    HematoxylinTextureGroup = TextureFeatureGroup(In, Nuclei)
-    HematoxylinGradientGroup = GradientFeatureGroup(In, Nuclei)
-    # calculate eosin features
-    EosinIntensityGroup = IntensityFeatureGroup(Ic, Cytoplasms)
-    EosinTextureGroup = TextureFeatureGroup(Ic, Cytoplasms)
-    EosinGradientGroup = GradientFeatureGroup(Ic, Cytoplasms)
+            np.logical_xor(Mask, binary_dilation(Nucleus, Disk))
+        )
```

Comment:
Maybe we need to create another file to be used in finding the region of cytoplasm?